File: expense-tracker/app/src/main/java/com/farhan/paisatrack/util/Format.kt
package com.farhan.paisatrack.util

import java.text.NumberFormat
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

object Format {
    private val inr: NumberFormat = NumberFormat.getCurrencyInstance(Locale("en", "IN"))
    private val dateFmt = DateTimeFormatter.ofPattern("dd MMM yyyy")
    private val dateTimeFmt = DateTimeFormatter.ofPattern("dd MMM yyyy, hh:mm a")
    private val timeFmt = DateTimeFormatter.ofPattern("hh:mm a")
    private val monthFmt = DateTimeFormatter.ofPattern("MMM yyyy")
    private val dayFmt = DateTimeFormatter.ofPattern("EEE, dd MMM")

    fun money(amount: Double): String = inr.format(amount)

    fun moneyShort(amount: Double): String {
        val a = kotlin.math.abs(amount)
        val sign = if (amount < 0) "-" else ""
        return when {
            a >= 10_000_000 -> "$sign₹%.2fCr".format(a / 10_000_000)
            a >= 100_000 -> "$sign₹%.2fL".format(a / 100_000)
            a >= 1_000 -> "$sign₹%.1fK".format(a / 1_000)
            else -> "$sign₹%.0f".format(a)
        }
    }

    private fun ldt(millis: Long) =
        Instant.ofEpochMilli(millis).atZone(ZoneId.systemDefault()).toLocalDateTime()

    fun date(millis: Long): String = ldt(millis).format(dateFmt)
    fun dateTime(millis: Long): String = ldt(millis).format(dateTimeFmt)
    fun time(millis: Long): String = ldt(millis).format(timeFmt)
    fun month(millis: Long): String = ldt(millis).format(monthFmt)
    fun day(millis: Long): String = ldt(millis).format(dayFmt)

    fun monthKey(millis: Long): String {
        val d = ldt(millis)
        return "%04d-%02d".format(d.year, d.monthValue)
    }
}

File: farhans-wallet-flutter/lib/core/utils/format.dart
import 'package:intl/intl.dart';

class Money {
  static final _inr = NumberFormat.currency(locale: 'en_IN', symbol: '₹', decimalDigits: 2);

  static String format(double amount) => _inr.format(amount);

  static String short(double amount) {
    final a = amount.abs();
    final sign = amount < 0 ? '-' : '';
    if (a >= 10000000) return '$sign₹${(a / 10000000).toStringAsFixed(2)}Cr';
    if (a >= 100000) return '$sign₹${(a / 100000).toStringAsFixed(2)}L';
    if (a >= 1000) return '$sign₹${(a / 1000).toStringAsFixed(1)}K';
    return '$sign₹${a.toStringAsFixed(0)}';
  }
}

class Dates {
  static final _date = DateFormat('dd MMM yyyy');
  static final _dateTime = DateFormat('dd MMM yyyy, hh:mm a');
  static final _time = DateFormat('hh:mm a');
  static final _month = DateFormat('MMM yyyy');
  static final _day = DateFormat('EEE, dd MMM');

  static DateTime _d(int ms) => DateTime.fromMillisecondsSinceEpoch(ms);

  static String date(int ms) => _date.format(_d(ms));
  static String dateTime(int ms) => _dateTime.format(_d(ms));
  static String time(int ms) => _time.format(_d(ms));
  static String month(int ms) => _month.format(_d(ms));
  static String day(int ms) => _day.format(_d(ms));

  static String monthKey(int ms) {
    final d = _d(ms);
    return '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}';
  }

  static String monthKeyNow() {
    final d = DateTime.now();
    return '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}';
  }
}
